File: main/detect-cycle-in-a-directed-graph/detect-cycle-in-a-directed-graph.py
'''Please note that it's Function problem i.e.
you need to write your solution in the form of Function(s) only.
Driver Code to call/invoke your function would be added by GfG's Online Judge.'''

from enum import Enum

# Your task is to complete this function
# Function should return True/False or 1/0
# Graph(graph) is a defaultict of type List
# n is no of Vertices
def isCyclic(n, graph):
    Color = Enum('Color', 'WHITE GRAY BLACK')
    vis = [Color.WHITE] * n

    def isCyclicFrom(u):
        if vis[u] is Color.WHITE:
            vis[u] = Color.GRAY
            if any(isCyclicFrom(v) for v in graph[u]):
                return True
            vis[u] = Color.BLACK
        elif vis[u] is Color.GRAY:
            return True
        elif vis[u] is Color.BLACK:
            return False
        else:
            raise AssertionError('internal error: invalid visitation state')

    return any(isCyclicFrom(u) for u in range(n))

File: main/detect-cycle-in-a-directed-graph/detect-cycle-in-a-directed-graph.cpp
/*
Please note that it's Function problem i.e.
you need to write your solution in the form of Function(s) only.
Driver Code to call/invoke your function would be added by GfG's Online Judge.*/


/*
The structure of the class is as follows
which contains an integer V denoting the no
of vertices and a list of adjacency vertices.
class Graph
{
    int V;
    list<int> *adj;
public :
    Graph(int V);
    void addEdge(int v,int w);
    bool isCyclic();
};
*/
/*You are required to complete this method*/
bool Graph::isCyclic()
{
    enum class Color { white, gray, black };
    vector<Color> vis (V);

    const function<bool(int)>
    isCyclicFrom {[this, &vis, &isCyclicFrom](const int v) {
        switch (vis[v]) {
        case Color::white:
            vis[v] = Color::gray;
            if (any_of(begin(adj[v]), end(adj[v]), isCyclicFrom)) return true;
            vis[v] = Color::black;
            return false;

        case Color::gray:
            return true;

        case Color::black:
            return false;
        }

        abort(); // unreachable
    }};

    for (auto v = 0; v != V; ++v)
        if (isCyclicFrom(v)) return true;

    return false;
}
